test/e2e/network/traffic_policy.go

```diff
@@ -0,0 +1,476 @@
+			tr.zonesByPod[pod.Name] = zone
+			tr.nodesByPod[pod.Name] = pod.Spec.NodeName
+		}
+	}
+
+	// use the largest zone to determine which node's port to use for nodeport endpoint requests
+	largestSeen := 0
+	for zone, nodes := range zonesWithNodes {
+		if len(nodes) > largestSeen {
+			largestSeen = len(nodes)
+			tr.endpointZone = zone
+		}
+	}
+	framework.ExpectNotEqual("", tr.endpointZone, "unable to find a zone with more than one node")
+
+	// always send to a node in the largest zone so balancing will result in more than 1 host name being seen
+	tr.endpointNode = zonesWithNodes[tr.endpointZone][0]
+	// same zone as endpointNode
+	peerNode := zonesWithNodes[tr.endpointZone][1]
+
+	// setup the endpoints which are used as target of curl in configurations below
+	// first determine the endpoint node IP (may be ip6)
+	family := v1.IPv4Protocol
+	if framework.TestContext.ClusterIsIPv6() {
+		family = v1.IPv6Protocol
+	}
+
+	ips := e2enode.GetAddressesByTypeAndFamily(tr.endpointNode, v1.NodeInternalIP, family)
+	framework.ExpectNotEqual(len(ips), 0)
+	nodeIp := ips[0]
+	nodePort := svc.Spec.Ports[0].NodePort
+
+	tr.nodeportEndpoint = net.JoinHostPort(nodeIp, strconv.Itoa(int(nodePort)))
+	tr.clusterIPEndpoint = svc.Spec.ClusterIP
+
+	var externalNode *v1.Node // node in a different zone from endpoint
+	for zone, nodes := range zonesWithNodes {
+		if zone != tr.endpointZone {
+			externalNode = nodes[0]
+			break
+		}
+	}
+
+	var internalClientNodeName string
+	var externalClientNodeName string
+	if topologyHints {
+		internalClientNodeName = tr.endpointNode.Name
+		externalClientNodeName = externalNode.Name
+	} else {
+		internalClientNodeName = tr.endpointNode.Name
+		externalClientNodeName = peerNode.Name
+	}
+	// create client pods
+	podClient := e2epod.NewPodClient(f)
+	tr.internalClientPod = createClientPod(podClient, tr.clusterIPEndpoint, f.Namespace.Name, internalClientNodeName, "internal")
+	tr.externalClientPod = createClientPod(podClient, tr.nodeportEndpoint, f.Namespace.Name, externalClientNodeName, "external")
+
+	return &tr
+}
+
+// This pod will repeatedly curl the specified endpoint and validate the result.
+// If expectedZone is specified, each request pod will be verified to reside in this zone. Otherwise the zone of the pod is not checked.
+// After a number of responses have been received checkResponses is used to validate which pods handled the request.
+func (tr *trafficPolicyTestResources) checkEndpointHost(c clientset.Interface, f *framework.Framework, endpoint string, expectedZone string, podName string, checkResponses seenPodsCheckFunc) {
+	var logs string
+	seenPods := make(map[string]string)
+	if pollErr := wait.Poll(5*time.Second, e2eservice.KubeProxyLagTimeout, func() (bool, error) {
+		var err error
+		logs, err = e2epod.GetPodLogs(c, f.Namespace.Name, podName, podName)
+		framework.ExpectNoError(err)
+		framework.Logf("Pod client logs: %s", logs)
+
+		logLines := strings.Split(logs, "\n")
+		if len(logLines) < 6 {
```

Comment:
What's the significance of this number of log lines here?

Reply:
I think this is just to wait to parse log lines until a meaningful number of responses have returned.

This was basically copied from the test/e2e/network/topology_hints.go which I used as a starting point.